### ISSUE TITLE
some error when building up toolchains

### ISSUE THREAD

**Issue by eunwoo** [bztsrc/raspi3-tutorial#47]
Hello,
I'm studying your tutorial. Thank you^^.
There's some error in building binutils.
Could you give me some help?
I followed your tutorial step by step.
I'm testing them by running Ubuntu 18.04.1 LTS in Virtual Box.

rpi@rpi-VirtualBox:~/binutils-2.30$ make -j4
make[1]: leaving directory '/home/rpi/binutils-2.30'
make[1]: nothing to do for 'all-target'
Configuring in ./etc
Configuring in ./intl
Configuring in ./isl
Configuring in ./libiberty
configure: error: cannot find sources () in .././isl
Makefile:5750: recipe for target 'configure-isl' failed
make[1]: *** [configure-isl] Error 1
make[1]: *** waiting for unended jobs....
configure: creating cache ./config.cache
checking whether make sets $(MAKE)... configure: creating cache ./config.cache
checking for a BSD-compatible install... /usr/bin/install -c
configure: creating cache ./config.cache
checking whether to enable maintainer-specific portions of Makefiles... no
checking for makeinfo... /home/rpi/binutils-2.30/missing makeinfo --split-size=5000000
configure: WARNING:
*** Makeinfo is missing. Info documentation will not be built.
checking for perl... perl
yes
checking for a BSD-compatible install... /usr/bin/install -c
checking whether NLS is requested... yes
checking for msgfmt... checking build system type... configure: updating cache ./config.cache
/usr/bin/msgfmt
checking for gmsgfmt... /usr/bin/msgfmt
configure: creating ./config.status
checking for xgettext... x86_64-pc-linux-gnu
checking host system type... /usr/bin/xgettext
x86_64-pc-linux-gnu
checking for x86_64-pc-linux-gnu-ar... ar
checking for x86_64-pc-linux-gnu-ranlib... ranlib
checking whether to install libiberty headers and static library... yes
configure: target_header_dir = libiberty
checking for x86_64-pc-linux-gnu-gcc... gcc
checking for msgmerge... /usr/bin/msgmerge
checking for x86_64-pc-linux-gnu-gcc... gcc
checking for C compiler default output file name... checking for C compiler default output file name... a.out
checking whether the C compiler works... a.out
checking whether the C compiler works... yes
checking whether we are cross compiling... no
checking for suffix of executables... yes
checking whether we are cross compiling... no
checking for suffix of executables... config.status: creating Makefile

checking for suffix of object files... 
checking for suffix of object files... o
checking whether we are using the GNU C compiler... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... yes
checking for gcc option to accept ISO C89... none needed
none needed
checking how to run the C preprocessor... checking build system type... x86_64-pc-linux-gnu
checking host system type... gcc -E
x86_64-pc-linux-gnu
checking for x86_64-pc-linux-gnu-ranlib... ranlib
checking for library containing strerror... checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... none required
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
yes
checking for an ANSI C-conforming const... checking for stdlib.h... yes
checking for inline... yes
inline
checking for string.h... checking for sys/types.h... yes
yes
checking for sys/stat.h... checking for memory.h... yes
yes
checking for stdlib.h... checking for strings.h... yes
yes
checking for string.h... checking for inttypes.h... yes
yes
checking for memory.h... checking for stdint.h... yes
yes
checking for strings.h... checking for unistd.h... yes
yes
checking minix/config.h usability... checking for inttypes.h... yes
no
checking minix/config.h presence... checking for stdint.h... no
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking for unistd.h... yes
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... no
checking whether gcc supports -W... yes
checking for off_t... yes
checking whether gcc supports -Wall... yes
checking whether gcc supports -Wwrite-strings... yes
yes
checking for size_t... checking whether gcc supports -Wc++-compat... yes
checking whether gcc supports -Wstrict-prototypes... yes
checking whether gcc supports -Wshadow=local... yes
checking for working alloca.h... yes
checking whether gcc supports -pedantic ... yes
checking whether gcc and cc understand -c and -o together... yes
checking for alloca... yes
checking for stdlib.h... (cached) yes
checking for unistd.h... (cached) yes
checking for getpagesize... yes
checking for an ANSI C-conforming const... yes
checking for working mmap... yes
checking for inline... inline
checking whether byte ordering is bigendian... yes
checking whether we are using the GNU C Library 2.1 or newer... yes
checking whether integer division by zero raises SIGFPE... no
checking for a BSD-compatible install... /usr/bin/install -c
checking for sys/file.h... yes
checking for sys/param.h... yes
checking for inttypes.h... yes
checking for limits.h... yes
yes
checking for stdint.h... checking for stdlib.h... (cached) yes
checking for malloc.h... yes
yes
checking for unsigned long long... checking for string.h... (cached) yes
checking for unistd.h... (cached) yes
checking for strings.h... (cached) yes
checking for sys/time.h... yes
checking for inttypes.h... yes
checking for time.h... yes
checking whether the inttypes.h PRIxNN macros are broken... yes
checking for sys/resource.h... yes
no
checking for ld used by GCC... ld
checking if the linker (ld) is GNU ld... yes
checking for shared library run path origin... checking for sys/stat.h... (cached) yes
checking for sys/mman.h... yes
done
checking for fcntl.h... checking argz.h usability... yes
checking for alloca.h... yes
checking for sys/pstat.h... yes
checking argz.h presence... yes
checking for argz.h... yes
no
checking for sys/sysmp.h... checking limits.h usability... no
checking for sys/sysinfo.h... yes
yes
checking limits.h presence... checking for machine/hal_sysinfo.h... yes
checking for limits.h... yes
no
checking locale.h usability... checking for sys/table.h... no
checking for sys/sysctl.h... yes
yes
checking locale.h presence... checking for sys/systemcfg.h... yes
checking for locale.h... yes
no
checking nl_types.h usability... checking for stdint.h... (cached) yes
checking for stdio_ext.h... yes
yes
checking nl_types.h presence... checking for process.h... yes
checking for nl_types.h... yes
no
checking malloc.h usability... checking for sys/prctl.h... yes
checking for sys/wait.h that is POSIX.1 compatible... yes
checking malloc.h presence... yes
checking whether time.h and sys/time.h may both be included... yes
checking for malloc.h... yes
checking stddef.h usability... yes
checking whether errno must be declared... yes
checking stddef.h presence... no
checking size of int... yes
checking for stddef.h... yes
checking for stdlib.h... (cached) yes
checking for string.h... (cached) yes
checking for unistd.h... (cached) yes
checking sys/param.h usability... 4
checking size of long... yes
checking sys/param.h presence... yes
checking for sys/param.h... yes
checking for feof_unlocked... 8
checking size of size_t... yes
checking for fgets_unlocked... yes
8
checking for getc_unlocked... checking for long long... yes
checking for getcwd... yes
checking size of long long... yes
checking for getegid... yes
8
checking for geteuid... checking for a 64-bit type... uint64_t
checking for intptr_t... yes
checking for getgid... yes
checking for getuid... yes
checking for uintptr_t... yes
checking for mempcpy... yes
checking for ssize_t... yes
checking for munmap... yes
checking for putenv... yes
checking for pid_t... yes
checking for setenv... yes
yes
checking for library containing strerror... checking for setlocale... none required
checking for asprintf... yes
checking for stpcpy... yes
checking for atexit... yes
checking for strcasecmp... yes
checking for basename... yes
checking for strdup... yes
checking for bcmp... yes
checking for strtoul... yes
checking for bcopy... yes
checking for tsearch... yes
yes
checking for bsearch... checking for __argz_count... yes
yes
checking for bzero... checking for __argz_stringify... yes
yes
checking for __argz_next... checking for calloc... yes
yes
checking for __fsetlocking... checking for clock... yes
yes
checking for iconv... checking for ffs... yes
checking for iconv declaration... yes
checking for getcwd... install-shextern size_t iconv (iconv_t cd, char * *inbuf, size_t *inbytesleft, char * *outbuf, size_t *outbytesleft);
checking for nl_langinfo and CODESET... yes
checking for getpagesize... yes
checking for LC_MESSAGES... yes
yes
checking for bison... no
checking whether NLS is requested... yes
checking whether included gettext is requested... no
checking for GNU gettext in libc... checking for gettimeofday... yes
yes
checking whether to use NLS... yes
checking where the gettext function comes from... libc
checking for aclocal... no
checking for autoconf... no
checking for autoheader... no
checking for index... configure: updating cache ./config.cache
configure: creating ./config.status
yes
checking for insque... yes
checking for memchr... yes
config.status: creating Makefile
checking for memcmp... config.status: creating config.intl
config.status: creating config.h
yes
config.status: executing default-1 commands
checking for memcpy... yes
checking for memmem... yes
checking for memmove... yes
checking for mempcpy... yes
checking for memset... yes
checking for mkstemps... yes
checking for putenv... yes
checking for random... yes
checking for rename... yes
checking for rindex... yes
checking for setenv... yes
checking for snprintf... yes
checking for sigsetmask... yes
checking for stpcpy... yes
checking for stpncpy... yes
checking for strcasecmp... yes
checking for strchr... yes
checking for strdup... yes
checking for strncasecmp... yes
checking for strndup... yes
checking for strnlen... yes
checking for strrchr... yes
checking for strstr... yes
checking for strtod... yes
checking for strtol... yes
checking for strtoul... yes
checking for strtoll... yes
checking for strtoull... yes
checking for strverscmp... yes
checking for tmpnam... yes
checking for vasprintf... yes
checking for vfprintf... yes
checking for vprintf... yes
checking for vsnprintf... yes
checking for vsprintf... yes
checking for waitpid... yes
checking for setproctitle... no
checking whether alloca needs Cray hooks... no
checking stack direction for C alloca... -1
checking for vfork.h... no
checking for fork... yes
checking for vfork... yes
checking for working fork... yes
checking for working vfork... (cached) yes
checking for _doprnt... no
checking for sys_errlist... yes
checking for sys_nerr... yes
checking for sys_siglist... yes
checking for external symbol _system_configuration... no
checking for __fsetlocking... yes
checking for canonicalize_file_name... yes
checking for dup3... yes
checking for getrlimit... yes
checking for getrusage... yes
checking for getsysinfo... no
checking for gettimeofday... (cached) yes
checking for on_exit... yes
checking for psignal... yes
checking for pstat_getdynamic... no
checking for pstat_getstatic... no
checking for realpath... yes
checking for setrlimit... yes
checking for sbrk... yes
checking for spawnve... no
checking for spawnvpe... no
checking for strerror... yes
checking for strsignal... yes
checking for sysconf... yes
checking for sysctl... yes
checking for sysmp... no
checking for table... no
checking for times... yes
checking for wait3... yes
checking for wait4... yes
checking whether basename is declared... yes
checking whether ffs is declared... yes
checking whether asprintf is declared... yes
checking whether vasprintf is declared... yes
checking whether snprintf is declared... yes
checking whether vsnprintf is declared... yes
checking whether calloc is declared... yes
checking whether getenv is declared... yes
checking whether getopt is declared... yes
checking whether malloc is declared... yes
checking whether realloc is declared... yes
checking whether sbrk is declared... yes
checking whether strtol is declared... yes
checking whether strtoul is declared... yes
checking whether strtoll is declared... yes
checking whether strtoull is declared... yes
checking whether strverscmp is declared... yes
checking whether strnlen is declared... yes
checking whether canonicalize_file_name must be declared... no
checking for stdlib.h... (cached) yes
checking for unistd.h... (cached) yes
checking for getpagesize... (cached) yes
checking for working mmap... yes
checking for working strncmp... yes
configure: updating cache ./config.cache
configure: creating ./config.status
config.status: creating Makefile
config.status: creating testsuite/Makefile
config.status: creating config.h
config.status: executing default commands
make[1]: leaving directory '/home/rpi/binutils-2.30' 
Makefile:850: recipe for target 'all' failed
make: *** [all] Error 2
rpi@rpi-VirtualBox:~/binutils-2.30$ 

This is the some part related to isl in Makefile.
.PHONY: maybe-maintainer-clean-isl maintainer-clean-isl
maybe-maintainer-clean-isl:
maybe-maintainer-clean-isl: maintainer-clean-isl

maintainer-clean-isl: 
	@[ -f ./isl/Makefile ] || exit 0; \
	r=`${PWD_COMMAND}`; export r; \
	s=`cd $(srcdir); ${PWD_COMMAND}`; export s; \
	$(HOST_EXPORTS) \
	for flag in $(EXTRA_HOST_FLAGS) V=1; do \
	  eval `echo "$$flag" | sed -e "s|^\([^=]*\)=\(.*\)|\1='\2'; export \1|"`; \
	done; \
	echo "Doing maintainer-clean in isl"; \
	(cd $(HOST_SUBDIR)/isl && \
	  $(MAKE) $(BASE_FLAGS_TO_PASS) "AR=$${AR}" "AS=$${AS}" \
	          "CC=$${CC}" "CXX=$${CXX}" "LD=$${LD}" "NM=$${NM}" \
	          "RANLIB=$${RANLIB}" \
	          "DLLTOOL=$${DLLTOOL}" "WINDRES=$${WINDRES}" "WINDMC=$${WINDMC}" \
	          maintainer-clean) \
	  || exit 1

2. when compiling gcc, some errors.

rpi@rpi-VirtualBox:~/gcc-8.1.0$ sudo make install-gcc
[sudo] rpi의 암호: 
/bin/bash ./mkinstalldirs /usr/local/cross-compiler /usr/local/cross-compiler
make[1]: entering '/home/rpi/gcc-8.1.0/host-x86_64-pc-linux-gnu/fixincludes' 
rm -rf /usr/local/cross-compiler/libexec/gcc/aarch64-elf/8.1.0/install-tools
/bin/bash ../.././fixincludes/../mkinstalldirs /usr/local/cross-compiler/libexec/gcc/aarch64-elf/8.1.0/install-tools
mkdir -p -- /usr/local/cross-compiler/libexec/gcc/aarch64-elf/8.1.0/install-tools
/bin/bash ../.././fixincludes/../mkinstalldirs /usr/local/cross-compiler/lib/gcc/aarch64-elf/8.1.0/install-tools/include
/usr/bin/install -c -m 644 ../.././fixincludes/README-fixinc \
  /usr/local/cross-compiler/lib/gcc/aarch64-elf/8.1.0/install-tools/include/README
/usr/bin/install -c fixinc.sh /usr/local/cross-compiler/libexec/gcc/aarch64-elf/8.1.0/install-tools/fixinc.sh
/usr/bin/install -c fixincl /usr/local/cross-compiler/libexec/gcc/aarch64-elf/8.1.0/install-tools/fixincl
/usr/bin/install -c mkheaders /usr/local/cross-compiler/libexec/gcc/aarch64-elf/8.1.0/install-tools/mkheaders
make[1]: leaving '/home/rpi/gcc-8.1.0/host-x86_64-pc-linux-gnu/fixincludes' 
make[1]: entering '/home/rpi/gcc-8.1.0/host-x86_64-pc-linux-gnu/lto-plugin'
make[2]: 디렉터리 '/home/rpi/gcc-8.1.0/host-x86_64-pc-linux-gnu/lto-plugin' 들어감
 /bin/mkdir -p '/usr/local/cross-compiler/libexec/gcc/aarch64-elf/8.1.0'
 /bin/bash ./libtool --tag=disable-static  --mode=install /usr/bin/install -c   liblto_plugin.la '/usr/local/cross-compiler/libexec/gcc/aarch64-elf/8.1.0'
libtool: install: /usr/bin/install -c .libs/liblto_plugin.so.0.0.0 /usr/local/cross-compiler/libexec/gcc/aarch64-elf/8.1.0/liblto_plugin.so.0.0.0
libtool: install: (cd /usr/local/cross-compiler/libexec/gcc/aarch64-elf/8.1.0 && { ln -s -f liblto_plugin.so.0.0.0 liblto_plugin.so.0 || { rm -f liblto_plugin.so.0 && ln -s liblto_plugin.so.0.0.0 liblto_plugin.so.0; }; })
libtool: install: (cd /usr/local/cross-compiler/libexec/gcc/aarch64-elf/8.1.0 && { ln -s -f liblto_plugin.so.0.0.0 liblto_plugin.so || { rm -f liblto_plugin.so && ln -s liblto_plugin.so.0.0.0 liblto_plugin.so; }; })
libtool: install: /usr/bin/install -c .libs/liblto_plugin.lai /usr/local/cross-compiler/libexec/gcc/aarch64-elf/8.1.0/liblto_plugin.la
libtool: finish: PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin:/sbin" ldconfig -n /usr/local/cross-compiler/libexec/gcc/aarch64-elf/8.1.0
----------------------------------------------------------------------
Libraries have been installed in:
   /usr/local/cross-compiler/libexec/gcc/aarch64-elf/8.1.0

If you ever happen to want to link against installed libraries
in a given directory, LIBDIR, you must either use libtool, and
specify the full pathname of the library, or use the `-LLIBDIR'
flag during linking and do at least one of the following:
   - add LIBDIR to the `LD_LIBRARY_PATH' environment variable
     during execution
   - add LIBDIR to the `LD_RUN_PATH' environment variable
     during linking
   - use the `-Wl,-rpath -Wl,LIBDIR' linker flag
   - have your system administrator add LIBDIR to `/etc/ld.so.conf'

See any operating system documentation about shared libraries for
more information, such as the ld(1) and ld.so(8) manual pages.
----------------------------------------------------------------------
make[2]: nothing to do for 'install-data-am'
make[2]: leaving '/home/rpi/gcc-8.1.0/host-x86_64-pc-linux-gnu/lto-plugin' 
make[1]: leaving '/home/rpi/gcc-8.1.0/host-x86_64-pc-linux-gnu/lto-plugin' 
make[1]: entering '/home/rpi/gcc-8.1.0/host-x86_64-pc-linux-gnu/gcc' 
make[1]: *** no rules for making target 'install'.  stop.
make[1]: leaving '/home/rpi/gcc-8.1.0/host-x86_64-pc-linux-gnu/gcc' 
Makefile:4342: recipe for target 'install-gcc' failed
make: *** [install-gcc] Error 2

**Comment by bztsrc** [comment]
Wrong forum. As you can read on the readme,
> IMPORTANT NOTE: this description is not about how to compile a cross-compiler in general. It's about how to compile one specifically for the aarch64-elf target. **If you have problems, google "how to build a gcc cross-compiler" or ask on an appropriate support forum for your operating system.** I can't and won't help you with your environment, you have to solve that on your own.

I suggest to ask your question on a gcc forum. But if I were you, I'd start by investigating the first error message in your ourput:

> configure: error: cannot find sources () in .././isl
> Makefile:5750: recipe for target 'configure-isl' failed

Cheers,
bzt

**Comment by adrianlopezroche** [comment]
Hi @bztsrc,

I think this is actually a bug in your instructions. The bug has to do with the link to `isl-0.18` under `binutils-2.30` (I'm using the same versions as in your tutorial). Specifically:

```
cd binutils-*
ln -s ../isl-* isl
```

will create a symlink named `isl` under `binutils-2.30`, but `isl` is not a real directory. Because it isn't a real directory, `isl/..` will point not to the parent of the symlink `isl` (namely, `binutils-2.30`) as might be expected, but to the parent of the directory `isl-0.18`. Thus, for a directory structure such as this one:

```
/path/to/sources/isl-0.18
/path/to/sources/binutils-2.30
```

`binutils-2.30/isl/..` will point to `/path/to/sources` rather than `binutils-2.30`. This is significant because the `Makefile` produced by `./configure` calls `cd ./isl` and expects `.././isl` to point to `isl` under `/path/to/sources/binutils-2.30` which it doesn't: it points to `/path/to/sources/./isl` instead, which doesn't exist and breaks the `Makefile` script. with the following error:
```
configure: error: cannot find sources () in .././isl
```

You can reproduce the effect with a simple experiment:

```
mkdir parent
cd parent
mkdir child1
mkdir child2
cd child2
ln -s ../child1 c1link
ls c1link/..
```
This will list the contents of `parent` rather than the contents of `child2`.

You've stated elsewhere that your example works on your system, but this is the reason it fails on mine and @eunwoo's. Are you using the exact versions of bintools and gcc listed in your tutorial?

**Comment by bztsrc** [comment]
Hi,

Yes, I've used binutils-2.30 and isl-0.18. But I haven't used that readme tbh (I already had the cross-compiler when I wrote that), and also I've merged a patch from someone else since, so there could be some error, although I really doubt that (the patch was mostly about checksums).

Here's my full cross-compiler compiling shell script that I've actually used (and I wrote the original readme after this, removing parts that are irrelevant for the tutorials, and I've unrolled a few loops for clearity):
```sh
#!/bin/bash

### Configuration
ARCHS="x86_64 aarch64"
DIR="/usr/local/gcc"
CFG=" --enable-shared --enable-threads=posix --enable-libmpx --with-system-zlib --with-isl --enable-__cxa_atexit \
--disable-libunwind-exceptions --enable-clocale=gnu --disable-libstdcxx-pch --disable-libssp --enable-plugin \
--disable-linker-build-id --enable-lto --enable-install-libiberty --with-linker-hash-style=gnu --with-gnu-ld\
--enable-gnu-indirect-function --disable-multilib --disable-werror --enable-checking=release --enable-default-pie \
--enable-default-ssp --enable-gnu-unique-object"

### Get binutils and gcc source. Modify versions in filenames as you like. Only place where versions referenced
[ ! -d tarballs ] && mkdir tarballs
echo "Downloading tarballs..."
cd tarballs
[ ! -f binutils-*tar* ] && wget http://ftpmirror.gnu.org/binutils/binutils-2.30.tar.gz
[ ! -f gcc-*tar* ] && wget http://ftpmirror.gnu.org/gcc/gcc-7.3.0/gcc-7.3.0.tar.gz
[ ! -f mpfr-*tar* ] && wget http://ftpmirror.gnu.org/mpfr/mpfr-4.0.1.tar.gz
[ ! -f gmp-*tar* ] && wget http://ftpmirror.gnu.org/gmp/gmp-6.1.2.tar.bz2
[ ! -f mpc-*tar* ] && wget http://ftpmirror.gnu.org/mpc/mpc-1.1.0.tar.gz
[ ! -f isl-*tar* ] && wget ftp://gcc.gnu.org/pub/gcc/infrastructure/isl-0.18.tar.bz2
[ ! -f cloog-*tar* ] && wget ftp://gcc.gnu.org/pub/gcc/infrastructure/cloog-0.18.1.tar.gz
cd ..

### Don't change anything below this line
echo -n "Unpacking tarballs... "
for i in tarballs/*.tar.gz; do d=${i%%.tar*};d=${d#*/}; echo -n "$d "; [ ! -d $d ] && tar -xzf $i; done
for i in tarballs/*.tar.bz2; do d=${i%%.tar*};d=${d#*/}; echo -n "$d "; [ ! -d $d ] && tar -xjf $i; done
echo "OK"

cd binutils-* && for i in isl; do ln -s ../$i-* $i 2>/dev/null || true; done && cd ..
cd gcc-* && for i in isl mpfr gmp mpc cloog; do ln -s ../$i-* $i 2>/dev/null || true; done && cd ..

#patch buggy gcc configure script to enable -fvisibility in gcc
cd gcc-*/gcc
cat configure|sed 's/gcc_cv_as_hidden=no/gcc_cv_as_hidden=yes/'|sed 's/gcc_cv_ld_hidden=no/gcc_cv_ld_hidden=yes/'>..c
mv ..c configure
chmod +x configure
cd ../..

### Compile gcc
for arch in $ARCHS; do
    echo "  -------------------------------------------- $arch -----------------------------------------------------"
    mkdir $arch-binutils 2>/dev/null
    cd $arch-binutils
    ../binutils-*/configure --prefix=$DIR --target=${arch}-elf $CFG
    make -j4
    make install
    cd ..
    mkdir $arch-gcc 2>/dev/null
    cd $arch-gcc
    ../gcc-*/configure --prefix=$DIR --target=${arch}-elf --enable-languages=c $CFG
    make -j4 all-gcc
    make install-gcc
    cd ..
done
```
I can still run this script without errors. I think the only difference is the gcc version, as it uses 7.3.0. Try my script, maybe that helps. Ubuntu by the way was reported to have broken gcc by several users, and some users have reported issues with compiling the sources too (interestingly also with problems related to isl). I'd suggest to use a different distro if you can, or ask around on Ubuntu forums. As a last resort, remove the symlink and copy the contents of isl-0.18 into binutils/isl and gcc/isl.

Good luck,
bzt

**Comment by adrianlopezroche** [comment]
Thank you. Looking at your script, I think I've spotted the reason why it works for you when it fails for us: you're creating a new directory for each target and compiling from there. The `configure-isl` rule for the `Makefile` generated by the bintools `configure` script will _not_ work when compiling directly inside `bintools-2.30`, but it works OK when compiling from a new directory (I won't bother going into why).

I haven't run your full script, but bintools-2.30 compiles successfully on Linux Mint (an Ubuntu-based distro) if I alter the steps of your tutorial (under "compiling the sources") as follows:

```
mkdir aarch64-binutils
cd aarch64-binutils
../binutils-*/configure --prefix=/usr/local/cross-compiler --target=aarch64-elf \
--enable-shared --enable-threads=posix --enable-libmpx --with-system-zlib --with-isl --enable-__cxa_atexit \
--disable-libunwind-exceptions --enable-clocale=gnu --disable-libstdcxx-pch --disable-libssp --enable-plugin \
--disable-linker-build-id --enable-lto --enable-install-libiberty --with-linker-hash-style=gnu --with-gnu-ld\
--enable-gnu-indirect-function --disable-multilib --disable-werror --enable-checking=release --enable-default-pie \
--enable-default-ssp --enable-gnu-unique-object
make -j4
make install
```

I haven't tried compiling gcc yet, but a similar edit is probably required.

**Comment by bztsrc** [comment]
> The configure-isl rule for the Makefile generated by the bintools configure script will not work when compiling directly inside bintools-2.30, but it works OK when compiling from a new directory

Wow, I've never thought of that not even in my wildest dreams! After all the '../isl-0.18' symlink in both binutils-2.30/isl and aarch64-binutils/isl should point to the same directory. This isl issue was puzzling for quite some time now, so thank you for spotting that!

> I won't bother going into why

I can understand that :-)

I've updated the readme, and since my gcc compilation also uses a new directory (which works for me) I've added that too.

Big thanks,
bzt
